### PR TITLE
Separate CoreNodesUI into CoreNodeModels, CoreNodeModelsWpf

### PR DESCRIPTION
**Pull request by pboyer**
### Why?

CoreNodesUI is a monolithic library.  It should not be WPF-specific.  By making this change, the core node library can be safely executed under Mono.

In this pull request, I've separated CoreNodesUI into two parts:

* **CoreNodeModels** - `NodeModel`s part of the "Core" Dynamo library
* **CoreNodeModelsWpf** - WPF-specific UI customizations for CoreNodeModels.

Note that, CoreNodeModels retains it's output name of "DSCoreNodesUI.dll"  Should be fixed in this issue:  http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-5852

### How?  

Mostly, this was just moving code around.  However, this change required significant cleanup of the `Watch` node and `SelectionBase.`

#### Changes to `Watch`

The architecture of the `Watch` node was highly confusing.  A method on the `Watch` node, called `GetWatchNode` returned a `WatchViewModel` which was assigned as a child of the root `WatchViewModel` hosted on Watch.  

Now, all of this logic lives in `WatchNodeViewCustomization,` which is composed around the `Watch` node.  `Watch` now emits an `EvaluationComplete` event which can be used for subscribing to intermediate updates from `VMDataBridge,`  

#### Changes to `SelectionBase`

`SelectionBase` hosted a `DelegateCommand`.  Beyond being a `Model` layer construct, this required the Prism library in order to support a Model.  The command has been moved to `SelectionBaseNodeViewCustomization,`  

### Effects on Dynamo

This change should not change the behavior of Dynamo at all.  

### Tests

Most of this PR is code shuffling, but I've made sure all of the Watch and Selection tests still pass:

watch

selection

### Reviewers 

- [x] @ikeough

### CODE REVIEW

File: src/DynamoCoreWpf/Interfaces/IWatchHandler.cs [hunk 1]
```diff
@@ -19,15 +19,6 @@ namespace Dynamo.Interfaces
...
     {
-        /// <summary>
-        /// 
-        /// </summary>
-        /// <param name="value"></param>
-        /// <param name="core"></param>
-        /// <param name="tag"></param>
-        /// <param name="showRawData"></param>
-        /// <param name="callback"></param>
-        /// <returns></returns>
```
Comment (pboyer) [inline src/DynamoCoreWpf/Interfaces/IWatchHandler.cs]: Unused...

File: src/Libraries/CoreNodeModels/CoreNodeModels.csproj [hunk 2]
```diff
@@ -34,16 +34,10 @@
     <WarningLevel>4</WarningLevel>
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/CoreNodeModels.csproj]: Note the solution is renamed here.

File: src/Libraries/CoreNodeModels/CoreNodeModels.csproj [hunk 3]
```diff
@@ -140,24 +104,6 @@
...
-    </Page>
-    <Page Include="UI\ElementSelectionControl.xaml">
-      <Generator>MSBuild:Compile</Generator>
-      <SubType>Designer</SubType>
-    </Page>
-    <Page Include="UI\IntegerSliderSettingsControl.xaml">
-      <SubType>Designer</SubType>
-      <Generator>MSBuild:Compile</Generator>
-    </Page>
-  </ItemGroup>
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/CoreNodeModels.csproj]: All UI related information is removed from the CoreNodeModels project.

File: src/Libraries/CoreNodeModels/CreateList.cs [hunk 4]
```diff
@@ -3,20 +3,10 @@
...
 namespace DSCoreNodesUI
 {
-    public class CreateListNodeViewCustomization : VariableInputNodeViewCustomization, INodeViewCustomization<CreateList>
-    {
-        public void CustomizeView(CreateList model, Dynamo.Controls.NodeView nodeView)
-        {
-            base.CustomizeView(model, nodeView);
-        }
-    }
-
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/CreateList.cs]: Moved to CoreNodeModelsWpf/NodeViewCustomizations/CreateList.cs

File: src/Libraries/CoreNodeModels/Formula.cs [hunk 5]
```diff
@@ -18,7 +18,6 @@ namespace DSCoreNodesUI
     [NodeDescription("Evaluates mathematical formulas. Uses NCalc: http://ncalc.codeplex.com/")]
     [NodeCategory(BuiltinNodeCategories.CORE_SCRIPTING)]
     [IsDesignScriptCompatible]
-    //[NodeDeprecated]
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Formula.cs]: Superfluous.

File: src/Libraries/CoreNodeModels/Input/BaseTypes.cs [hunk 6]
```diff
@@ -162,7 +162,7 @@ public string Value
 
                 try
                 {
-                    _parsed = ParseValue(value, new[] { '\n' }, idList, _convertToken);
+                    _parsed = ParseValue(value, new[] { '\n' }, idList, ConvertToken);
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Input/BaseTypes.cs]: Protected member should be named properly.

File: src/Libraries/CoreNodeModels/Input/DoubleSlider.cs [hunk 7]
```diff
@@ -95,19 +94,25 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
 
         #endregion
 
+        private static double ConvertBack(string value)
+        {
+            double val = 0.0;
+            double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out val);
+            return val;
+        }
+
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Input/DoubleSlider.cs]: DoubleDisplay is an IValueConverter, an interface derived from PresentationCore (yuck).  Here it is inlined.

File: src/Libraries/CoreNodeModels/Input/IntegerSlider.cs [hunk 8]
```diff
@@ -57,19 +55,25 @@ public int Min
             }
         }
 
+        private static int ConvertBack(string value)
+        {
+            int val = 0;
+            int.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out val);
+            return val;
+        }
+
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Input/IntegerSlider.cs]: Similar to DoubleSlider.cs.

File: src/Libraries/CoreNodeModels/Input/String.cs [hunk 9]
```diff
@@ -17,8 +16,7 @@ protected override bool UpdateValueCore(string name, string value, UndoRedoRecor
         {
             if (name == "Value")
             {
-                var converter = new StringDisplay();
-                Value = converter.ConvertBack(value, typeof(string), null, null) as string;
+                Value = HttpUtility.HtmlEncode(value);
                 return true;
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Input/String.cs]: This is exactly the same logic as in StringDisplay.

File: src/Libraries/CoreNodeModels/Selection.cs [hunk 10]
```diff
@@ -62,8 +57,6 @@ private set
             }
         }
 
-        public DelegateCommand SelectCommand { get; set; }
-
         public string Prefix { get; set; }
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Selection.cs]: This is moved to the `NodeViewCustomization` class.

File: src/Libraries/CoreNodeModels/Selection.cs [hunk 11]
```diff
@@ -264,38 +256,4 @@ public virtual void UpdateSelection(IEnumerable<TSelection> newSelection)
...
-        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
-        {
-            return value.ToString();
-        }
-
-        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
-        {
-            return null;
-        }
-    }
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Selection.cs]: Moved to CoreNodeModelsWpf/Converters

File: src/Libraries/CoreNodeModels/Watch.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+namespace Dynamo.Nodes
+{
+    [NodeName("Watch")]
+    [NodeCategory(BuiltinNodeCategories.CORE_VIEW)]
+    [NodeDescription("Visualize the output of node. ")]
+    [NodeSearchTags("print", "output", "display")]
+    [IsDesignScriptCompatible]
+    public class Watch : NodeModel
+    {
+        public event Action<Object> EvaluationComplete;
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/Watch.cs]: Used in order to obtain UI updates mid-evaluation from VMDataBridge.

File: src/Libraries/CoreNodeModels/WatchImageCore.cs [hunk 13]
```diff
@@ -16,8 +15,6 @@ namespace Dynamo.Nodes
     [IsDesignScriptCompatible]
     public class WatchImageCore : NodeModel
     {
-        private Image image;
-
```
Comment (pboyer) [inline src/Libraries/CoreNodeModels/WatchImageCore.cs]: Out WindowsBase!  Damn you!

File: src/Libraries/CoreNodeModelsWpf/Controls/DoubleSliderSettingsControl.xaml.cs [hunk 14]
```diff
@@ -3,7 +3,7 @@
 using System.Windows.Data;
 using Dynamo.Controls;
 
-namespace DSCoreNodesUI
+namespace Dynamo.Wpf
```
Comment (pboyer) [inline src/Libraries/CoreNodeModelsWpf/Controls/DoubleSliderSettingsControl.xaml.cs]: Remove these horrific monolithic namespaces.

File: src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf.csproj [hunk 15]
```diff
@@ -0,0 +1,138 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ImportGroup Label="PropertySheets">
+    <Import Project="$(SolutionDir)Config/CS.props" />
+  </ImportGroup>
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
```
Comment (pboyer) [inline src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf.csproj]: All of the NodeViewCustomizations live here.

File: src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf.csproj [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    <Compile Include="Controls\IntegerSliderSettingsControl.xaml.cs">
+      <DependentUpon>IntegerSliderSettingsControl.xaml</DependentUpon>
+    </Compile>
+    <Compile Include="NodeViewCustomizations\SelectionBase.cs" />
+    <Compile Include="Converters\SelectionButtonContentConverter.cs" />
+    <Compile Include="Converters\SelectionToTextConverter.cs" />
+    <Compile Include="NodeViewCustomizations\StringInput.cs" />
+    <Compile Include="NodeViewCustomizations\WatchImage.cs" />
+    <Compile Include="NodeViewCustomizations\Watch.cs" />
+  </ItemGroup>
```
Comment (pboyer) [inline src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf.csproj]: All of the WPF-specific logic is here.

File: src/Libraries/CoreNodesUI/Watch.cs [hunk 17]
```diff
@@ -1,228 +0,0 @@
-﻿using System;
-using System.Collections.Generic;
-
-using Dynamo.Controls;
-using Dynamo.Core.Threading;
```
Comment (pboyer) [inline src/Libraries/CoreNodesUI/Watch.cs]: For reasons I don't understand, this did not get a proper diff, nor did WatchNodeViewCustomization.  It is so significantly modified that it probably wouldn't help...

File: src/Libraries/VMDataBridge/VMDataBridge.cs [hunk 18]
```diff
@@ -23,7 +23,11 @@ public static DataBridge Instance
             {
                 lock (mutex)
                 {
-                    return instance ?? (instance = new DataBridge());   
+                    if (instance == null)
+                    {
+                        instance = new DataBridge();
+                    }
+                    return instance;
                 }
```
Comment (pboyer) [inline src/Libraries/VMDataBridge/VMDataBridge.cs]: This was really difficult to debug.  This is just more readable.

File: test/DynamoCoreTests/WatchNodeTests.cs [hunk 19]
```diff
@@ -98,17 +119,17 @@ public void WatchLiterals()
...
+            var watchPoint = model.CurrentWorkspace.NodeFromWorkspace("f1581148-9318-40fa-9402-61557255162a") as Watch;
 
-            WatchViewModel node = watchNumber.GetWatchNode();
+            var node = GetWatchViewModel(watchNumber);
             Assert.AreEqual("5", node.NodeLabel);
 
-            node = watchBoolean.GetWatchNode();
-            Assert.AreEqual("False", node.NodeLabel);
+            node = GetWatchViewModel(watchBoolean);
+            Assert.AreEqual("false", node.NodeLabel);
```
Comment (pboyer) [inline test/DynamoCoreTests/WatchNodeTests.cs]: Believe it or not, this was the sole mistake that relegate this to the Failure category.

File: src/DynamoCore/Models/PortModel.cs [hunk 20]
```diff
@@ -23,12 +21,12 @@ public class PortModel : ModelBase
...
         /// Event triggered when a port is connected.
         /// </summary>
-        public event Action<ConnectorModel> PortConnected;
+        public event Action<PortModel, ConnectorModel> PortConnected;
 
         /// <summary>
         /// Event triggered when a port is disconnected.
         /// </summary>
-        public event PortConnectedHandler PortDisconnected;
+        public event Action<PortModel> PortDisconnected;
```
Comment (pboyer) [inline src/DynamoCore/Models/PortModel.cs]: This badly needed cleanup.  Now the `PortDisconnected` and `PortConnected` events are much more sensible and it removes the used, non-type safe args in `PortConnectedHandler`.

File: src/Libraries/VMDataBridge/VMDataBridge.cs (hunk 18 above)
Comment (ikeough) [inline src/Libraries/VMDataBridge/VMDataBridge.cs]: I'm glad I'm not the only one who thinks that is more readable.

**Comment by ikeough** [comment]
@pboyer When you fix the merge conflicts, LGTM. 

Also, it would be good if we had a doc somewhere on our wiki that outlined the dos and don'ts of writing code that was mono compliant. Even if that doc just told us to use the mono migration analyzer.

**Pull request merged**